druid/src/widget/svg.rs

```diff
@@ -0,0 +1,294 @@
+    /// Create an SVG-drawing widget from a valid SVG string literal.
+    ///
+    /// The SVG will scale to fit its box constraints.
+    /// If SVG is invalid a blank SVG will be rendered instead.
+    pub fn new_from_str(svg_str: &str) -> impl Widget<T> {
+        Svg {
+            tree: svg_tree_from_str(svg_str),
+            phantom: Default::default(),
+        }
+    }
+
+    /// Create an SVG-drawing widget from a valid path to an .svg file.
+    ///
+    /// The SVG will scale to fit its box constraints.
+    /// If SVG is missing or invalid a blank SVG will be rendered instead.
+    pub fn new_from_path(path: impl AsRef<std::path::Path>) -> impl Widget<T> {
+        Svg {
+            tree: svg_tree_from_path(path),
+            phantom: Default::default(),
+        }
+    }
+
+    /// Measure the SVG's size
+    #[allow(clippy::needless_return)]
+    fn get_size(&self) -> Size {
+        let root = self.tree.root();
+        match *root.borrow() {
+            usvg::NodeKind::Svg(svg) => {
+                return Size::new(svg.size.width(), svg.size.height());
+            }
+            _ => {
+                //TODO: I don't think this is reachable?
+                error!("This SVG has no size for some reason.");
+                return Size::ZERO;
+            }
+        };
+    }
+}
+
+#[cfg(feature = "svg")]
+impl<T: Data> Widget<T> for Svg<T> {
+    fn paint(&mut self, paint_ctx: &mut PaintCtx, base_state: &BaseState, _data: &T, _env: &Env) {
+        //TODO: options for aspect ratio or scaling based on height
+        let scale = base_state.size().width / self.get_size().width;
+
+        let origin_x = (base_state.size().width - (self.get_size().width * scale)) / 2.0;
+        let origin_y = (base_state.size().height - (self.get_size().height * scale)) / 2.0;
+        let origin = Point::new(origin_x, origin_y);
+
+        svg_to_piet(&self.tree, scale, origin, paint_ctx);
+    }
+
+    fn layout(
+        &mut self,
+        _layout_ctx: &mut LayoutCtx,
+        bc: &BoxConstraints,
+        _data: &T,
+        _env: &Env,
+    ) -> Size {
+        bc.debug_check("SVG");
+
+        if bc.is_width_bounded() {
+            bc.max()
+        } else {
+            bc.constrain(self.get_size())
+        }
+    }
+
+    fn event(&mut self, _ctx: &mut EventCtx, _event: &Event, _data: &mut T, _env: &Env) {}
+
+    fn update(&mut self, _ctx: &mut UpdateCtx, _old_data: Option<&T>, _data: &T, _env: &Env) {}
+}
+
+// Create a raw usvg tree from an SVG string.
+// Pass this tree to `svg_to_piet()` to draw it.
+#[cfg(feature = "svg")]
+pub fn svg_tree_from_str(svg_str: &str) -> usvg::Tree {
+    let re_opt = usvg::Options {
+        keep_named_groups: false,
+        ..usvg::Options::default()
+    };
+
+    match usvg::Tree::from_str(svg_str, &re_opt) {
+        Ok(tree) => tree,
+        Err(err) => {
+            error!("{}", err);
+            usvg::Tree::from_str(EMPTY_SVG, &re_opt).unwrap()
+        }
+    }
```

Comment:
This needs to return a `Result` rather than trying to handle the error internally.